### PR TITLE
Total Scattering improvements

### LINKED ISSUE #32446: Fix unit conversion in TotScatCalculateSelfScattering
The workflow algorithm TotScatCalculateSelfScattering currently generates a "per detector" correction with units of TOF and then focusses this data (still in TOF). Having discussed with the ORNL team we think the data should be converted to dSpacing before focussing. 

I have reviewed the history of the TotScatCalculateSelfScattering algorithm and the workflow was originally set up in the ISIS Powder diffraction scripts for Polaris. A unit conversion to dSpacing was in an early version of the workflow (in the form of a call to AlignDetectors) but was removed - it seems like this was a mistake:

https://github.com/mantidproject/mantid/commit/619ffc7f84a66b4a656f7c2a58bd8cc720eac0fa#diff-2e03bd9793088ae4e583ae7f3fa770cb10414b67f5c0fcff9535fcd9ae26027c

So this change is to reinstate the dSpacing unit conversion (which can now be done in ConvertUnits). 

I'll also fix a glitch at the end of v1 of CalculatePlaczekSelfScattering where the output unit is set to match the x unit of the input workspace (which is typically wavelength) even though the x data has been calculated in TOF. There's a choice here: a) set the output unit to be TOF, b) calculate the x values in the x unit of the input workspace.

I think we need to do a check on how much this moves the result for a POLARIS input workspace. If it's a big change we should get in touch with Helen Playford to let her know. She worked on validating the output of TotScatCalculateSelfScattering when it was being implemented

### PR DESCRIPTION
**Pull request by DanielMurphy22**
**Description of work.**
- Improved CalculatePlaczekSelfScattering version 1 to output in whatever units the input workspace has, added validation that the IncidentSpecta is in units of Wavelength
- Altered TotScattCalculateSelfScattering algorithm to run DiffractionFoccussing in Q and actually use CalculatePlaczekSelfScattering version

**To test:**

Check the current main branch behaviour first: 
I added the command:
`daniel_ssc = mantid.CloneWorkspace(InputWorkspace=self_scattering_correction, OutputWorkspace="daniel_ssc")`
to output the self_scattering_correction workspace on [Line97 of polaris_alg.py](https://github.com/mantidproject/mantid/blob/main/scripts/Diffraction/isis_powder/polaris_routines/polaris_algs.py#L97)
Then run this script (ask me if I need to email data):
```
from mantid.simpleapi import *
import numpy as np
from isis_powder import polaris, SampleDetails
import time

start_time = time.time()
config_file_path = r"/home/danielmurphy/Downloads/polaris-calculate-pdf/polaris_config_example.yaml"
polaris = polaris.Polaris(config_file=config_file_path, user_name="test", mode="PDF")

sample_details = SampleDetails(height=4.0, radius=0.2985, center=[0, 0, 0], shape='cylinder')
sample_details.set_material(chemical_formula='Si')
polaris.set_sample_details(sample=sample_details)

polaris.create_vanadium(first_cycle_run_no="98532", multiple_scattering=False)
polaris.focus(run_number="98533", input_mode='Summed')
polaris.create_total_scattering_pdf(run_number="98533", merge_banks=False)
print("Script run time", time.time() - start_time)
```
The current main code gives these plots:
Screenshot 2021-10-12 at 14 40 43

Next please, check out my changes, make sure you still have the CloneWorkspace command and have built. The new result is as so:
Screenshot 2021-10-12 at 14 30 15

I have found that there is only a negligible difference (10E-16 for values ~ 0.3) in the output of the old and new CalculatePlaczekSelfScattering v1 algorithms. There is a noticeable difference in the PDF output from TotScattCalcSelfScattering due to focussing in Q instead of TOF. I think the differences are small or in unimportant places that they don't seem to have much of an effect on the final PDF. See the images below for some of my comparisons.

Screenshot 2021-10-12 at 14 58 24

Above: The self scattering correction workspaces (rebinning gave the vertical lines!) from the CloneWorkspace command, focused in TOF (main) focused in Q (this PR branch) and on the left their difference. The big features seem to be where the old and new ways give different end values for each spectrum.
Below are zoomed in plots to display the extent of the differences
Screenshot 2021-10-12 at 15 00 09
Below show the differences for the focused_Q and PDF_R workspaces which are output from the create_pdf call in the script at the top of these instructions (again with zoomed in versions).
Screenshot 2021-10-12 at 14 49 39

Fixes #32446 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by DannyHindson** [comment]
Thanks for all the plots! I've started having a look through all of this and will continue in the morning. The fact there are differences in the daniels_ssc plots looks about right - I would have expected the extra ConvertUnits call before focussing to change this result. If that doesn't actually cause a massive change in the final output then that's fine. I don't think I had much intuition about its eventual impact on the pdf but seemed like the correction itself should move a bit.

**Review by DannyHindson — CHANGES_REQUESTED** [review]

**Review by DannyHindson — CHANGES_REQUESTED** [review]
I've added a few more comments but they're mostly simple\small things now.

The PyExec method in TotScatCalculateSelfScattering is getting to the point where it might be worth pulling out some of the raw code bits (feg or loops) out into separate functions. But since we may well end up changing the workflow around a bit more with the "per detector" correction approach it's probably not worth polishing this yet.

File: Framework/Algorithms/src/CalculatePlaczekSelfScattering.cpp
Comment (DanielMurphy22) [reply Framework/Algorithms/src/CalculatePlaczekSelfScattering.cpp:165]: I think this means I can remove all of the else statement below?

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/TotScatCalculateSelfScattering.py
Comment (DannyHindson) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/TotScatCalculateSelfScattering.py:94]: This calculation for the rebin parameters looks fine now. If we were pursuing this approach of focussing the correction I think it would be better if we calculated different rebin parameters per group and then called GroupDetectors per group. DiffractionFocussing calculates rebin parameters per group (but with the unwanted side effect that it always uses log bins). Not sure it's worth you doing this work now though. So please could you add a comment in here saying "TO DO: calculate rebin parameters per group" or similar

I noticed this because the q range of the correction for bank 1 is quite small (0 to ~30 only) so when you rebin using a common set of rebin parameters (spanning a q range of ~0-375) across the whole workspace you end up with quite a lot of zeroes for the bank 1 correction for the high q values (in daniel_ssc). The focussed correction eventually gets rebinned to match the focused data in polaris_algs so doesn't ultimately have a huge effect but perhaps reduces the granularity of the correction for bank 1 unnecessarily

**Review by DannyHindson — CHANGES_REQUESTED** [review]
I've added a couple of v small further requests.

Here's what I get for the correction after these changes:

image

As discussed there's a gap in the overall workflow where the focused ws that is being corrected needs to be absolutely normalized. So the fact the correction is bigger than the "signal" isn't itself a concern. I'm just going to spend a bit of time checking why the correction happens to almost equal "1" - that's perhaps slightly suspicious. 
The other obvious characteristic of the plot is that the correction for each spectrum drops to zero above a particular Q value. This is just a consequence of using a common set of rebin parameters for the whole workspace before the call to GroupDetectors so that makes sense.

File: Framework/Algorithms/src/CalculatePlaczekSelfScattering.cpp
Comment (DanielMurphy22) [reply Framework/Algorithms/src/CalculatePlaczekSelfScattering.cpp:149]: I think I should also remove the setYUnit(""); line above?

Comment (DannyHindson) [reply Framework/Algorithms/src/CalculatePlaczekSelfScattering.cpp:149]: I think you probably need to keep this line. I think the workspace has YUnit="Counts" and YUnitLabel="" at the point it's created. So I think the YUnit does need blanking out so that both YUnit and YUnitLabel are blank strings.

Comment (DanielMurphy22) [reply Framework/Algorithms/src/CalculatePlaczekSelfScattering.cpp:149]: Does ConvertUnits figure out the input Unit as Wavelength in some other way?

**Comment by DanielMurphy22** [comment]
Currently blocked by pre-commit on Jenkins: https://github.com/mantidproject/mantid/pull/32931

**Comment by DannyHindson** [comment]
@KedoKudo - we discussed some changes to the v1 CalculatePlaczekSelfScattering algorithm a few weeks ago. This PR contains the changes. Here's a summary of what we've done. You may want to do similar changes to v2 but I'll leave that up to you:

- the algorithm now outputs the correction in the same units as the input workspace. It previously always output in TOF
- the change in the previous bullet point has allowed us to simplify\remove some code relating to unit conversions. It's no longer necessary to create Kernel::Units::Wavelength and Kernel::Units::Tof objects and we don't need to call specInfo.getDetectorValues any more
- we don't do setDistribution(False) any more at the end. I think this feature was in place because at ISIS we calculate the correction per detector and then "focus" the detector. The DiffractionFocussing algorithm only works on histogram data so required distribution to be False. I think we're going to stop focussing the correction like this so the output now has distribution=True
- we've removed a line that added +1 to the correction. I think this was introduced originally because it meant that you ended up with S(Q)-1 when you subtracted the self scattering correction but we now do the "-1" explicitly in the calling code. I think this is less obscure

The PR contains some other changes to our workflow that calls CalculatePlaczekSelfScattering that you can ignore

**Comment by KedoKudo** [comment]
> @KedoKudo - we discussed some changes to the v1 CalculatePlaczekSelfScattering algorithm a few weeks ago. This PR contains the changes. Here's a summary of what we've done. You may want to do similar changes to v2 but I'll leave that up to you:
> 
> * the algorithm now outputs the correction in the same units as the input workspace. It previously always output in TOF
> * the change in the previous bullet point has allowed us to simplify\remove some code relating to unit conversions. It's no longer necessary to create Kernel::Units::Wavelength and Kernel::Units::Tof objects and we don't need to call specInfo.getDetectorValues any more
> * we don't do setDistribution(False) any more at the end. I think this feature was in place because at ISIS we calculate the correction per detector and then "focus" the detector. The DiffractionFocussing algorithm only works on histogram data so required distribution to be False. I think we're going to stop focussing the correction like this so the output now has distribution=True
> * we've removed a line that added +1 to the correction. I think this was introduced originally because it meant that you ended up with S(Q)-1 when you subtracted the self scattering correction but we now do the "-1" explicitly in the calling code. I think this is less obscure
> 
> The PR contains some other changes to our workflow that calls CalculatePlaczekSelfScattering that you can ignore

Thanks for the headsup. 

@peterfpeterson @Kvieta1990, could you take a look at the changes here to see if we need to mirror the changes?  I don't see anything critical that needs to happen immediately, but I could also be wrong

**Comment by Kvieta1990** [comment]
@KedoKudo @peterfpeterson The v1 Placzek correction is implemented in MTS and I am still doing a test for MTS where we now have just put in everything (including the v1 Placzek correction). Before we worry about the v2 of Placzek, I just want to have a robustly working version of MTS first. Only after that will we consider implementing the v2 Placzek. I will put this down into my notes and after I finish my systematic testing with real data for MTS, we can come back to worry about this.

Does that sound good?

**Comment by DannyHindson** [comment]
> @KedoKudo @peterfpeterson The v1 Placzek correction is implemented in MTS and I am still doing a test for MTS where we now have just put in everything (including the v1 Placzek correction). Before we worry about the v2 of Placzek, I just want to have a robustly working version of MTS first. Only after that will we consider implementing the v2 Placzek. I will put this down into my notes and after I finish my systematic testing with real data for MTS, we can come back to worry about this.
> 
> Does that sound good?

I didn't realise until now that you're using the v1 Placzek correction at ORNL. I was aware Chen was working on a v2 algorithm but I'd previously thought that v2 represented your first use of Mantid for this correction. I guess we'd better slow down a bit on this while we discuss because you probably don't want v1 to change during or after your testing. When do you expect to finish your testing that includes the v1 Placzek correction?

We should discuss the v1 changes at some point and see if they're helpful for you. The main change we've made that I think will affect you is the removal of a "+1" from the correction (4th bullet in my earlier message).

**Comment by Kvieta1990** [comment]
> > @KedoKudo @peterfpeterson The v1 Placzek correction is implemented in MTS and I am still doing a test for MTS where we now have just put in everything (including the v1 Placzek correction). Before we worry about the v2 of Placzek, I just want to have a robustly working version of MTS first. Only after that will we consider implementing the v2 Placzek. I will put this down into my notes and after I finish my systematic testing with real data for MTS, we can come back to worry about this.
> > Does that sound good?
> 
> I didn't realise until now that you're using the v1 Placzek correction at ORNL. I was aware Chen was working on a v2 algorithm but I'd previously thought that v2 represented your first use of Mantid for this correction. I guess we'd better slow down a bit on this while we discuss because you probably don't want v1 to change during or after your testing. When do you expect to finish your testing that includes the v1 Placzek correction?
> 
> We should discuss the v1 changes at some point and see if they're helpful for you. The main change we've made that I think will affect you is the removal of a "+1" from the correction (4th bullet in my earlier message).

Sorry, double checking our implementation of Placzek correction in `mantidtotalscattering`, I found that we were using our own version of Placzek correction within `mantidtotalscattering` so I guess the change on your side won't influence our workflow, at least for `mantidtotalscattering`. Other than that, I am not sure. @DannyHindson @KedoKudo @peterfpeterson

**Comment by DannyHindson** [comment]
@Kvieta1990 @peterfpeterson - thanks for the update. Since you're not using v1 after all we'll go ahead with these changes. As discussed, you can consider later on whether you want to replicate any of these changes in the v2 algorithm once you get to the point of integrating v2 in your workflow. I'll leave that up to the ORNL team

**Review by DannyHindson — APPROVED** [review]

**Comment by peterfpeterson** [comment]
It is my understanding that the second order requires the first order, to some extant, so we will eventually need to track down the cause of the difference. I'm fine with this being merged.

**Pull request merged**
